### DISCUSSION TITLE
Markdoc design

### DISCUSSION THREAD

**Discussion by adriangalilea** [markdoc/markdoc#486] [General]
Let me start by saying that I deeply feel the need for something like markdoc to exist, [code and content should be separated](https://stephango.com/file-over-app). So, that leaves `.mdx` out of the question and sadly `.md` is not enough.

With that in mind, I would like to share my thoughts:

# Why `{% %} {% / %}`?

One of the appeals for markdoc is allow authoring to non technical people, and seemingly trivial/minor decisions like these that create more boilerplate and uncomfortable key combinations, can be bothersome in day to day use, specially for newbies.

Markdoc
```
{% callout type="warning" %}
Some text.
{% /callout %}
```

Some alternatives:

[mdc](https://content.nuxt.com/usage/markdown#block-components)
```
::alert{type="warning"}
Some text.
::
```

I'd personally choose `<>` like mdx but without imports and code separation...
```
<Alert type="warning">
Some text.
</Alert>
```

So is there a reason for `{% %} {% / %}`?

Is this something that could be changed?

# Why not an official file extension?

[Why remain ambiguous?](https://stephango.com/style)

For instance, Astro uses `.mdoc`, while other implementations use `.md`, this makes the "framework" less recognisable and confusing, [making adoption harder](https://www.anca.io/posts/dropping-astro-markdoc-for-the-super-smooth-astro-markdown-obsidian-setup).

I think Markdoc would benefit from an official file extension.

I'm aware that these are subjective points and perhaps uncomfortable to discuss, but I bring them in good faith.

Let me know what you think.

Thank you,

Adrian.

**Comment by mfix-stripe** [comment]
Hi @adriangalilea, want to address the second point. `.mdoc` is our official file extension. Since Markdoc is a superset of Markdown, it supports Markdown out of the box, and thus is why some folks choose to use the `.md` file extension for their documents, and get things like editor support (Markdown syntax highlighting etc.) for free. 

As support for Markdoc grows, I expect to see a a greater use of `.mdoc` in the future.

**Reply by adriangalilea** [reply]
> Hi @adriangalilea, want to address the second point. `.mdoc` is our official file extension. Since Markdoc is a superset of Markdown, it supports Markdown out of the box, and thus is why some folks choose to use the `.md` file extension for their documents, and get things like editor support (Markdown syntax highlighting etc.) for free. 
> 
> 
> 
> As support for Markdoc grows, I expect to see a a greater use of `.mdoc` in the future. 

Hmm, I explicitly looked for this info in the docs and couldn't find it, I think this should be very obvious and upfront, best on the front page or on the first link, at worst on the FAQ, also I've skimmed through a bunch of official repos and examples and I see mostly .md files.

![image](https://github.com/markdoc/markdoc/assets/90320947/4e401d94-df39-42fb-a1a3-33262e76062c)

What gives?

**Comment by rpaul-stripe** [comment]
On the question of syntax, there are a few reasons we ended up with `{% tag %}`.

We chose not to use conventional HTML-like tags (`<tag>`) because Markdown processors typically have very specific behavior that they use for processing HTML and we didn't want to create conflicts with that. Some Markdoc users still rely on embedded HTML [via this tokenizer technique](https://gist.github.com/rpaul-stripe/941eb22c4779ea87b1adf7715d76ca08) and use it for various purposes.

We decided against things like `::tag` in mdc or asciidoc's [delimiter block syntax](https://docs.asciidoctor.org/asciidoc/latest/blocks/delimited/#nesting) because we wanted to require that the opening and closing syntax have explicit, matched names. This helps identify nesting errors and makes it significantly easier to troubleshoot structural issues in the document.

The original Markdoc prototype used a bbcode-like syntax with square brackets (like this: `[tag]example[/tag]`) which I chose for simplicity. The problem we had with this syntax is that there were inline cases that were ambiguous with reference-style links (e.g. it's not clear if `[foo][bar]` is a link or a pair of opening tags). In order to avoid that issue, we switched to the curly brace, so for much of Markdoc's early life the tags looked like this: `{tag}`.

When I added support for tags inside of fenced code blocks, there were a bunch of cases where it was difficult to distinguish between tags and regular content in the fences due to the ubiquity of curly braces in so many programming languages. We initially debated making it so that there was a separate tag syntax just for fenced code blocks, but ultimately decided that we wanted to have the tag syntax remain consistent everywhere. We had to pick a syntax that would be mostly unambiguous when mixed with conventional source code in mainstream languages.

After considering several options, we settled on `{%`, which has a fair amount of precedent given that it's also the syntax used in Django's templating system and the Liquid templating language.

We chose not to make the syntax configurable because we wanted to encourage consistency between Markdoc projects and avoid adding unnecessary complexity to tools that parse and analyze Markdoc content. I think we are unlikely to change it at this point.
